README
NAME
    AnyEvent::Filesys::Notify - An AnyEvent compatible module to monitor
    files/directories for changes

VERSION
    version 0.23_10

SYNOPSIS
        use AnyEvent::Filesys::Notify;

        my $notifier = AnyEvent::Filesys::Notify->new(
            dirs     => [ qw( this_dir that_dir ) ],
            interval => 2.0,             # Optional depending on underlying watcher
            filter   => sub { shift !~ /\.(swp|tmp)$/ },
            cb       => sub {
                my (@events) = @_;
                # ... process @events ...
            },
        );

        # enter an event loop, see AnyEvent documentation
        Event::loop();

DESCRIPTION
    This module provides a cross platform interface to monitor files and
    directories within an AnyEvent event loop. The heavy lifting is done by
    Linux::INotify2 or Mac::FSEvents on their respective O/S. A fallback
    which scans the directories at regular intervals is include for other
    systems. See "IMPLEMENTATIONS" for more on the backends.

    Events are passed to the callback (specified as a CodeRef to "cb" in the
    constructor) in the form of AnyEvent::Filesys::Notify::Events.

METHODS
  new()
    A constructor for a new AnyEvent watcher that will monitor the files in
    the given directories and execute a callback when a modification is
    detected. No action is take until a event loop is entered.

    Arguments for new are:

    dirs
            dirs => [ '/var/log', '/etc' ],

        An ArrayRef of directories to watch. Required.

    interval
            interval => 1.5,   # seconds

        Specifies the time in fractional seconds between file system checks
        for the AnyEvent::Filesys::Notify::Role::Fallback implementation.

        Specifies the latency for Mac::FSEvents for the
        "AnyEvent::Filesys::Notify::Role::FSEvents" implementation.

        Ignored for the "AnyEvent::Filesys::Notify::Role::Inotify2"
        implementation.

    filter
            filter => qr/\.(ya?ml|co?nf|jso?n)$/,
            filter => sub { shift !~ /\.(swp|tmp)$/,

        A CodeRef or Regexp which is used to filter wanted/unwanted events.
        If this is a Regexp, we attempt to match the absolute path name and
        filter out any that do not match. If a CodeRef, the absolute path
        name is passed as the only argument and the event is fired only if
        there sub returns a true value.

    cb
            cb  => sub { my @events = @_; ... },

        A CodeRef that is called when a modification to the monitored
        directory(ies) is detected. The callback is passed a list of
        AnyEvent::Filesys::Notify::Events. Required.

    backend
            backend => 'Fallback',
            backend => 'FreeBSD',
            backend => '+My::Filesys::Notify::Role::Backend',

        Force the use of the specified backend. The backend is assumed to
        have the "AnyEvent::Filesys::Notify::Role" prefix, but you can force
        a fully qualified name by prefixing it with a plus. Optional.

    no_external
            no_external => 1,

        This is retained for backward compatibility. Using "backend ="
        'Fallback'> is preferred. Force the use of the "Fallback" watcher
        implementation. This is not encouraged as the "Fallback" implement
        is very inefficient, but it does not require either Linux::INotify2
        nor Mac::FSEvents. Optional.

WATCHER IMPLEMENTATIONS
  INotify2 (Linux)
    Uses Linux::INotify2 to monitor directories. Sets up an "AnyEvent->io"
    watcher to monitor the "$inotify->fileno" filehandle.

  FSEvents (Mac)
    Uses Mac::FSEvents to monitor directories. Sets up an "AnyEvent->io"
    watcher to monitor the "$fsevent->watch" filehandle.

  KQueue (FreeBSD/Mac)
    Uses IO::KQueue to monitor directories. Sets up an "AnyEvent->io"
    watcher to monitor the "IO::KQueue" object.

    WARNING - IO::KQueue and the "kqueue()" system call require an open
    filehandle for every directory and file that is being watched. This
    makes it impossible to watch large directory structures (and inefficient
    to watch moderately sized directories). The use of the KQueue backend is
    discouraged.

  Fallback
    A simple scan of the watched directories at regular intervals. Sets up
    an "AnyEvent->timer" watcher which is executed every "interval" seconds
    (or fractions thereof). "interval" can be specified in the constructor
    to AnyEvent::Filesys::Notify and defaults to 2.0 seconds.

    This is a very inefficient implementation. Use one of the others if
    possible.

Why Another Module For File System Notifications
    At the time of writing there were several very nice modules that
    accomplish the task of watching files or directories and providing
    notifications about changes. Two of which offer a unified interface that
    work on any system: Filesys::Notify::Simple and File::ChangeNotify.

    AnyEvent::Filesys::Notify exists because I need a way to simply tie the
    functionality those modules provide into an event framework. Neither of
    the existing modules seem to work with well with an event loop.
    Filesys::Notify::Simple does not supply a non-blocking interface and
    File::ChangeNotify requires you to poll an method for new events. You
    could fork off a process to run Filesys::Notify::Simple and use an event
    handler to watch for notices from that child, or setup a timer to check
    File::ChangeNotify at regular intervals, but both of those approaches
    seem inefficient or overly complex. Particularly, since the underlying
    watcher implementations (Mac::FSEvents and Linux::INotify2) provide a
    filehandle that you can use and IO event to watch.

    This is not slight against the authors of those modules. Both are well
    respected, are certainly finer coders than I am, and built modules which
    are perfect for many situations. If one of their modules will work for
    you by all means use it, but if you are already using an event loop,
    this module may fit the bill.

SEE ALSO
    Modules used to implement this module AnyEvent, Mac::FSEvents,
    Linux::INotify2, Moose.

    Alternatives to this module Filesys::Notify::Simple, File::ChangeNotify.

BUGS
    Please report any bugs or suggestions at <http://rt.cpan.org/>

    Forcing the "IO::KQueue" backend on a Mac does not seem to work. The
    "IO::KQueue" backend seems to be working fine on FreeBSD. I don't have
    the experience or time to fix it on a Mac. I would greatly appreciate
    any help troubleshooting this.

CONTRIBUTORS
    Thanks to Gasol Wu <gasol.wu@gmail.com> who contributed the FreeBSD
    support for IO::KQueue.

AUTHOR
    Mark Grimes, <mgrimes@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Mark Grimes, <mgrimes@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changes for AnyEvent::Filesys::Notify

1.22 - Tue Dec  5 16:47:23 EST 2017
    Adds new directories to watch *before* searching them for new entities.
    - When using `parse_events => 1` with inotify, there is a race condition
      where entities in a new directory could be missed. We need to add the
      new directory to the watch list before searching it for new files or
      sub-directories.
    - Reported with PR by @ufobat (GH#11). Thanks!
    - Upgrade is recommended on Linux

    Adds skip_dirs support as contributed by @kyzn

    Other minor changes:
    - Improves the diagnostic when Linux::Inotify2->new fails
    - Adds skip_dirs support as contributed by @kyzn
    - Remove 5.21 from travis build (doesn't exist)
    - Removing Mac::FSEvents in META.json

1.21 - Thu Jun 23 15:13:59 EDT 2016
    Updates KQueue to remove file handles from watch list for deleted paths
    - Addresses issue with leaking file handles. Thanks to dave@jetcafe.org.
      (RT#115562).
    - Upgrade is recommended if you use the KQueue backend.

1.20 - Fri Jun 17 13:59:05 EDT 2016
    Runs bug test against KQueue if possible; improves comments
    - Ensures files are added to watch before calling in callback (RT#115380)
      Files that were modified in the callback were sometimes not added to the
      watch list first (specifically in KQueue and Inotify2). This patch
      ensures that files/directories are added to the watch before the callback
      is called. Thanks to Dave Hayes <dave@jetcafe.org> for the bug report.
    - If you modified _process_events (a private method) this could break your
      code.
    - Upgrade is recommended if you use the KQueue or Inotify2 backends.

1.19 - Tue May 26 15:20:03 EDT 2015
    - Converts utility functions to methods to support subclassing (RT#104714).
      The use of local functions (rather than methods) makes subclassing
      more difficult. This change converts all utility functions in Notify.pm
      to methods. Thanks to Dave Hayes for submitting RT#104714.
    - Any users attempting to subclass should consider upgrading.

1.18 - Wed May 20 20:27:58 EDT 2015
    - Adds version to some files which were missing it
    - No change in functionality. No need to upgrade.

1.17 - Wed May 20 08:07:04 EDT 2015
    - Switches from namespace::sweep to namespace::autoclean (RT#104554).
      autoclean no longer depends on Class::MOP (a heavy prereq).
      Thanks to Karen Etheridge (ether++) for report.
    - No change in functionality. No need to upgrade.

1.16 - Fri Feb 27 13:38:46 EST 2015
    - Rewrites the test functions to accommodate optional events
      Parsing events from Inotify2 (instead of re-scanning) results in
      extraneous modifications in some situations. Tests shouldn't fail if
      these optional events are triggered.

1.15 - Fri Aug 15 20:14:38 EDT 2014
    - Adds minimum version of perl to the prereqs

1.12 - Sun Dec 22 14:04:39 EST 2013
    - Uses KQueue on any BSD RT#91556

1.11 - Wed Dec 11 08:59:54 EST 2013
    - Uses IO::KQueue on all freebsd-like systems (ie, gnufreebsd).
    - Fixes minor POD linking problem.

1.10 - Tue Dec 10 11:42:06 EST 2013
    - If the parse_events flag is set, use events from Inotify2 instead of
      rescanning all files (RT#82847).  Thanks to Carsten Wolff
      <carsten@wolffcarsten.de> for the patch.
    - Generates 'created' events for files in a new subdirs when parsing events.
      This ensures that we don't miss any.
    - Refactors the test routines to improve reliability and increase code
      reuse.

0.24 - Fri Oct  4 11:44:42 EDT 2013
    - Makes an official release of after converting to Moo. CPANTesters didn't
      seem to have any issues with the conversion. Hopefully this change will
      be seemless for users.

0.23_10 - Mon Sep 30 20:06:50 EDT 2013
    EXPERIMENTAL DEVELOPERS RELEASE

    - Attempts to convert from Moose to Moo for our OO infrastructure. We
      don't need much of Moose's power, so let's take advantage of Moo's speed.
    - Adds version requirements to the prereqs. Earlier version of some of
      the prereq seemed to produce an unstable result. Current versions seem
      fine.
    - Fixes Changes file to adhere to CPAN::Changes spec
    - Replaces namespace::autoclean with namespace::sweep as
      namespace::autoclean relies on Class::MOP

0.22 - Tue Mar  5 14:15:44 EST 2013
    - This is an important release for KQueue users. Prior versions had several
      bugs in the KQueue backend. As a result, KQueue was effectively polling
      the filesystem with no delay. KQueue has a serious limitation which makes
      it impractical for monitoring medium to large sets of files (namely, you
      need an open filehandle for each directory AND file that you are
      monitoring). The use of the KQueue backend is discouraged.

    - Fixes KQueue backend bug which causes CPU usage to peg at 100%
      (thank to gdxhsw++ for the bug report and patch)
    - Updates the documentation and issue warnings for KQueue filehandles
    - Fixes errors in documentation
    - Adds namespace::autoclean to the Event object

0.21 - Mon Feb  4 15:52:47 EST 2013
    - Converts from File::Find::Rule to Path::Interator::Rule
      - Should speed things up marginally.
      - Adds benchmark and profiling tools
    - Marks tests as TODO if we are on gnufreebsd or gnukfreebsd
      - Based on diag output, tests appear to work, but we get failures from
        cpantesters. I don't have access to this platform to investigate
        further.

0.19 - Mon Jan 14 20:30:38 EST 2013
    - Fixes KQueue bugs.  This works on a Mac, hoping it will work on FreeBSD.

0.18 - Thu Jan 10 10:39:22 EST 2013
    - Fixes non-mac bug where 'plan' is called twice.
      Thanks to syohex for the patch.

0.17 - Wed Jan  9 19:53:53 EST 2013
    - Renames backend roles to reflect the implementation rather than the O/S
      (Inotify2, FSEvents and KQueue)

lib/AnyEvent/Filesys/Notify.pm
package AnyEvent::Filesys::Notify;

# ABSTRACT: An AnyEvent compatible module to monitor files/directories for changes

use Moo;
use Moo::Role ();
use MooX::late;
use namespace::autoclean;
use AnyEvent;
use Path::Iterator::Rule;
use Cwd qw/abs_path/;
use AnyEvent::Filesys::Notify::Event;
use Carp;
use Try::Tiny;

our $VERSION = '1.22';
my $AEFN = 'AnyEvent::Filesys::Notify';

has dirs         => ( is => 'ro', isa => 'ArrayRef[Str]', required => 1 );
has cb           => ( is => 'rw', isa => 'CodeRef',       required => 1 );
has interval     => ( is => 'ro', isa => 'Num',           default  => 2 );
has no_external  => ( is => 'ro', isa => 'Bool',          default  => 0 );
has backend      => ( is => 'ro', isa => 'Str',           default  => '' );
has filter       => ( is => 'rw', isa => 'RegexpRef|CodeRef' );
has parse_events => ( is => 'rw', isa => 'Bool',          default  => 0 );
has skip_subdirs => ( is => 'ro', isa => 'Bool',          default  => 0 );
has _fs_monitor  => ( is => 'rw', );
has _old_fs => ( is => 'rw', isa => 'HashRef' );
has _watcher => ( is => 'rw', );

sub BUILD {
    my $self = shift;

    $self->_old_fs( $self->_scan_fs( $self->dirs ) );

    $self->_load_backend;
    return $self->_init;    # initialize the backend
}

sub _process_events {
    my ( $self, @raw_events ) = @_;

    # Some implementations provided enough information to parse the raw events,
    # other require rescanning the file system (ie, Mac::FSEvents).
    # The original behavior was to rescan in all implementations, so we
    # have added a flag to avoid breaking old code.

    my @events;

    if ( $self->parse_events and $self->can('_parse_events') ) {
        @events =
          $self->_parse_events( sub { $self->_apply_filter(@_) }, @raw_events );
    } else {
        my $new_fs = $self->_scan_fs( $self->dirs );
        @events =
          $self->_apply_filter( $self->_diff_fs( $self->_old_fs, $new_fs ) );
        $self->_old_fs($new_fs);

        # Some backends (when not using parse_events) need to add files
        # (KQueue) or directories (Inotify2) to the watch list after they are
        # created. Give them a chance to do that here.
        $self->_post_process_events(@events)
          if $self->can('_post_process_events');
    }

    $self->cb->(@events) if @events;

    return \@events;
}

sub _apply_filter {
    my ( $self, @events ) = @_;

    if ( ref $self->filter eq 'CODE' ) {
        my $cb = $self->filter;
        @events = grep { $cb->( $_->path ) } @events;
    } elsif ( ref $self->filter eq 'Regexp' ) {
        my $re = $self->filter;
        @events = grep { $_->path =~ $re } @events;
    }

    return @events;
}

# Return a hash ref representing all the files and stats in @path.
# Keys are absolute path and values are path/mtime/size/is_dir
# Takes either array or arrayref
sub _scan_fs {
    my ( $self, @args ) = @_;

    # Accept either an array of dirs or a array ref of dirs
    my @paths = ref $args[0] eq 'ARRAY' ? @{ $args[0] } : @args;

    my $fs_stats = {};

    my $rule = Path::Iterator::Rule->new;
    $rule->skip_subdirs(qr/./)
        if (ref $self) =~ /^AnyEvent::Filesys::Notify/
        && $self->skip_subdirs;
    my $next = $rule->iter(@paths);
    while ( my $file = $next->() ) {
        my $stat = $self->_stat($file)
          or next; # Skip files that we can't stat (ie, broken symlinks on ext4)
        $fs_stats->{ abs_path($file) } = $stat;
    }

    return $fs_stats;
}

sub _diff_fs {
    my ( $self, $old_fs, $new_fs ) = @_;
    my @events = ();

    for my $path ( keys %$old_fs ) {
        if ( not exists $new_fs->{$path} ) {
            push @events,
              AnyEvent::Filesys::Notify::Event->new(
                path   => $path,
                type   => 'deleted',
                is_dir => $old_fs->{$path}->{is_dir},
              );
        } elsif (
            $self->_is_path_modified( $old_fs->{$path}, $new_fs->{$path} ) )
        {
            push @events,
              AnyEvent::Filesys::Notify::Event->new(
                path   => $path,
                type   => 'modified',
                is_dir => $old_fs->{$path}->{is_dir},
              );
        }
    }

    for my $path ( keys %$new_fs ) {
        if ( not exists $old_fs->{$path} ) {
            push @events,
              AnyEvent::Filesys::Notify::Event->new(
                path   => $path,
                type   => 'created',
                is_dir => $new_fs->{$path}->{is_dir},
              );
        }
    }

    return @events;
}

sub _is_path_modified {
    my ( $self, $old_path, $new_path ) = @_;

    return 1 if $new_path->{mode} != $old_path->{mode};
    return   if $new_path->{is_dir};
    return 1 if $new_path->{mtime} != $old_path->{mtime};
    return 1 if $new_path->{size} != $old_path->{size};
    return;
}

# Originally taken from Filesys::Notify::Simple --Thanks Miyagawa
sub _stat {
    my ( $self, $path ) = @_;

    my @stat = stat $path;

    # Return undefined if no stats can be retrieved, as it happens with broken
    # symlinks (at least under ext4).
    return undef unless @stat;

    return {
        path   => $path,
        mtime  => $stat[9],
        size   => $stat[7],
        mode   => $stat[2],
        is_dir => -d _,
    };

}

# Figure out which backend to use:
# I would prefer this to be done at compile time not object build, but I also
# want the user to be able to force the Fallback role. Something like an
# import flag would be great, but Moose creates an import sub for us and
# I'm not sure how to cleanly do it. Maybe need to use traits, but the
# documentation suggests traits are for application of roles by object.
# This will work for now.
sub _load_backend {
    my $self = shift;

    if ( $self->backend ) {

        # Use the AEFN::Role prefix unless the backend starts with a +
        my $prefix  = "${AEFN}::Role::";
        my $backend = $self->backend;
        $backend = $prefix . $backend unless $backend =~ s{^\+}{};

        try { Moo::Role->apply_roles_to_object( $self, $backend ); }
        catch {
            croak "Unable to load the specified backend ($backend). You may "
              . "need to install Linux::INotify2, Mac::FSEvents or IO::KQueue:"
              . "\n$_";
        }
    } elsif ( $self->no_external ) {
        Moo::Role->apply_roles_to_object( $self, "${AEFN}::Role::Fallback" );
    } elsif ( $^O eq 'linux' ) {
        try {
            Moo::Role->apply_roles_to_object( $self,
                "${AEFN}::Role::Inotify2" );
        }
        catch {
            croak "Unable to load the Linux plugin. You may want to install "
              . "Linux::INotify2 or specify 'no_external' (but that is very "
              . "inefficient):\n$_";
        }
    } elsif ( $^O eq 'darwin' ) {
        try {
            Moo::Role->apply_roles_to_object( $self,
                "${AEFN}::Role::FSEvents" );
        }
        catch {
            croak "Unable to load the Mac plugin. You may want to install "
              . "Mac::FSEvents or specify 'no_external' (but that is very "
              . "inefficient):\n$_";
        }
    } elsif ( $^O =~ /bsd/ ) {
        try {
            Moo::Role->apply_roles_to_object( $self, "${AEFN}::Role::KQueue" );
        }
        catch {
            croak "Unable to load the BSD plugin. You may want to install "
              . "IO::KQueue or specify 'no_external' (but that is very "
              . "inefficient):\n$_";
        }
    } else {
        Moo::Role->apply_roles_to_object( $self, "${AEFN}::Role::Fallback" );
    }

    return 1;
}

1;

__END__

=pod

=head1 NAME

AnyEvent::Filesys::Notify - An AnyEvent compatible module to monitor files/directories for changes

=head1 VERSION

version 1.22

=head1 STATUS

=for html <img src="https://travis-ci.org/mvgrimes/AnyEvent-Filesys-Notify.svg?branch=master" alt="Build Status">
<a href="https://metacpan.org/pod/AnyEvent::Filesys::Notify"><img alt="CPAN version" src="https://badge.fury.io/pl/AnyEvent-Filesys-Notify.svg" /></a>

=head1 SYNOPSIS

    use AnyEvent::Filesys::Notify;

    my $notifier = AnyEvent::Filesys::Notify->new(
        dirs     => [ qw( this_dir that_dir ) ],
        interval => 2.0,             # Optional depending on underlying watcher
        filter   => sub { shift !~ /\.(swp|tmp)$/ },
        cb       => sub {
            my (@events) = @_;
            # ... process @events ...
        },
        parse_events => 1,  # Improves efficiency on certain platforms
    );

    # enter an event loop, see AnyEvent documentation
    Event::loop();

=head1 DESCRIPTION

This module provides a cross platform interface to monitor files and
directories within an L<AnyEvent> event loop. The heavy lifting is done by
L<Linux::INotify2> or L<Mac::FSEvents> on their respective O/S. A fallback
which scans the directories at regular intervals is include for other systems.
See L</WATCHER IMPLEMENTATIONS> for more on the backends.

Events are passed to the callback (specified as a CodeRef to C<cb> in the
constructor) in the form of L<AnyEvent::Filesys::Notify::Event>s.

=head1 METHODS

=head2 new()

A constructor for a new AnyEvent watcher that will monitor the files in the
given directories and execute a callback when a modification is detected. 
No action is take until a event loop is entered.

Arguments for new are:

=over 4

=item dirs 

    dirs => [ '/var/log', '/etc' ],

An ArrayRef of directories to watch. Required.

=item interval

    interval => 1.5,   # seconds

Specifies the time in fractional seconds between file system checks for
the L<AnyEvent::Filesys::Notify::Role::Fallback> implementation.

Specifies the latency for L<Mac::FSEvents> for the
C<AnyEvent::Filesys::Notify::Role::FSEvents> implementation.

Ignored for the C<AnyEvent::Filesys::Notify::Role::Inotify2> implementation.

=item filter

    filter => qr/\.(ya?ml|co?nf|jso?n)$/,
    filter => sub { shift !~ /\.(swp|tmp)$/,

A CodeRef or Regexp which is used to filter wanted/unwanted events. If this
is a Regexp, we attempt to match the absolute path name and filter out any
that do not match. If a CodeRef, the absolute path name is passed as the
only argument and the event is fired only if there sub returns a true value.

=item cb

    cb  => sub { my @events = @_; ... },

A CodeRef that is called when a modification to the monitored directory(ies) is
detected. The callback is passed a list of
L<AnyEvent::Filesys::Notify::Event>s. Required.

=item backend

    backend => 'Fallback',
    backend => 'KQueue',
    backend => '+My::Filesys::Notify::Role::Backend',

Force the use of the specified backend. The backend is assumed to have the
C<AnyEvent::Filesys::Notify::Role> prefix, but you can force a fully qualified
name by prefixing it with a plus. Optional.

=item no_external

    no_external => 1,

This is retained for backward compatibility. Using C<backend => 'Fallback'>
is preferred. Force the use of the L</Fallback> watcher implementation. This is
not encouraged as the L</Fallback> implement is very inefficient, but it does
not require either L<Linux::INotify2> nor L<Mac::FSEvents>. Optional.

=item parse_events

    parse_events => 1,

In backends that support it (currently INotify2), parse the events instead of
rescanning file system for changed C<stat()> information. Note, that this might
cause slight changes in behavior. In particular, the Inotify2 backend will
generate an additional 'modified' event when a file changes (once when opened
for write, and once when modified).

=item skip_subdirs

    skip_subdirs => 1,

Skips subdirectories and anything in them while building a list of files/dirs
to watch. Optional.

=back

=head1 WATCHER IMPLEMENTATIONS

=head2 INotify2 (Linux)

Uses L<Linux::INotify2> to monitor directories. Sets up an C<AnyEvent-E<gt>io>
watcher to monitor the C<$inotify-E<gt>fileno> filehandle.

=head2 FSEvents (Mac)

Uses L<Mac::FSEvents> to monitor directories. Sets up an C<AnyEvent-E<gt>io>
watcher to monitor the C<$fsevent-E<gt>watch> filehandle.

=head2 KQueue (BSD/Mac)

Uses L<IO::KQueue> to monitor directories. Sets up an C<AnyEvent-E<gt>io>
watcher to monitor the C<IO::KQueue> object.

B<WARNING> - L<IO::KQueue> and the C<kqueue()> system call require an open
filehandle for every directory and file that is being watched. This makes
it impossible to watch large directory structures (and inefficient to watch
moderately sized directories). The use of the KQueue backend is discouraged.

=head2 Fallback

A simple scan of the watched directories at regular intervals. Sets up an
C<AnyEvent-E<gt>timer> watcher which is executed every C<interval> seconds
(or fractions thereof). C<interval> can be specified in the constructor to
L<AnyEvent::Filesys::Notify> and defaults to 2.0 seconds.

This is a very inefficient implementation. Use one of the others if possible.

=head1 Why Another Module For File System Notifications

At the time of writing there were several very nice modules that accomplish
the task of watching files or directories and providing notifications about
changes. Two of which offer a unified interface that work on any system:
L<Filesys::Notify::Simple> and L<File::ChangeNotify>.

L<AnyEvent::Filesys::Notify> exists because I need a way to simply tie the
functionality those modules provide into an event framework. Neither of the
existing modules seem to work with well with an event loop.
L<Filesys::Notify::Simple> does not supply a non-blocking interface and
L<File::ChangeNotify> requires you to poll an method for new events. You could
fork off a process to run L<Filesys::Notify::Simple> and use an event handler
to watch for notices from that child, or setup a timer to check
L<File::ChangeNotify> at regular intervals, but both of those approaches seem
inefficient or overly complex. Particularly, since the underlying watcher
implementations (L<Mac::FSEvents> and L<Linux::INotify2>) provide a filehandle
that you can use and IO event to watch.

This is not slight against the authors of those modules. Both are well 
respected, are certainly finer coders than I am, and built modules which 
are perfect for many situations. If one of their modules will work for you
by all means use it, but if you are already using an event loop, this
module may fit the bill.

=head1 SEE ALSO

Modules used to implement this module L<AnyEvent>, L<Mac::FSEvents>,
L<Linux::INotify2>, L<Moose>.

Alternatives to this module L<Filesys::Notify::Simple>, L<File::ChangeNotify>.

=head1 CONTRIBUTORS

=for stopwords Gasol Wu E<lt>gasol.wu@gmail.comE<gt> who contributed the BSD support for IO::KQueue Dave Hayes E<lt>dave@jetcafe.orgE<gt> Carsten Wolff E<lt>carsten@wolffcarsten.deE<gt>

=over 4

=item *

Gasol Wu E<lt>gasol.wu@gmail.comE<gt> who contributed the BSD support for IO::KQueue

=item *

Dave Hayes E<lt>dave@jetcafe.orgE<gt>

=item *

Carsten Wolff E<lt>carsten@wolffcarsten.deE<gt>

=back

=head1 AUTHOR

Mark Grimes, E<lt>mgrimes@cpan.orgE<gt>

=head1 SOURCE

Source repository is at L<https://github.com/mvgrimes/AnyEvent-Filesys-Notify>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<http://github.com/mvgrimes/AnyEvent-Filesys-Notify/issues>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Mark Grimes, E<lt>mgrimes@cpan.orgE<gt>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
